tools/build_langserver/langserver/analyzer.go

```diff
@@ -1,3 +1,199 @@
+
+// Analyzer is a wrapper around asp.Analyzer
+// This is being loaded into a handler on initialization
+type Analyzer struct {
+	parser *asp.Parser
+	BuiltIns map[string]*RuleDef
+}
+
+// RuleDef is a wrapper around asp.FuncDef,
+// it also includes a Header(function definition)
+// And Argument map stores the name and the information of the arguments this rule has
+type RuleDef struct {
+	*asp.FuncDef
+	Header string
+	ArgMap map[string]*Argument
+}
+
+// Argument is a wrapper around asp.Argument,
+// this is used to store the argument information for specific rules,
+// and it also tells you if the argument is required
+type Argument struct {
+	*asp.Argument
+	definition string
+	required bool
+}
+
+// Identifier is a wrapper around asp.Identifier
+// Including the starting line and the ending line number
+type Identifier struct {
+	*asp.IdentStatement
+	StartLine int
+	EndLine int
+}
+
+func newAnalyzer() *Analyzer {
+	state := core.NewDefaultBuildState()
+	parser := parse.GetParserWithBuiltins(state)
+
+	a := &Analyzer{
+		parser:parser,
+	}
+	a.BuiltInsRules()
+
+	return a
+}
+
+// BuiltInsRules gets all the builtin functions and rules as a map, and store it in Analyzer.BuiltIns
+func (a *Analyzer) BuiltInsRules() {
+	statementMap := make(map[string]*RuleDef)
+
+	for _, statement := range a.parser.GetAllBuiltinStatements() {
+		// Saves FuncDefs into the statementMap if it's a None private rule
+		if statement.FuncDef != nil && !strings.HasPrefix(statement.FuncDef.Name, "_") {
+			statementMap[statement.FuncDef.Name] = newRuleDef(statement.FuncDef)
+		}
+	}
+	a.BuiltIns = statementMap
+}
+
+// IdentFromPos gets the Identifier given a lsp.Position
+func (a *Analyzer) IdentFromPos(uri lsp.DocumentURI, position lsp.Position, filecontent []string) (*Identifier, error) {
+	idents, err := a.IdentFromFile(uri, filecontent)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ident := range idents {
+		if position.Line >= ident.StartLine && position.Line <= ident.EndLine {
+			return ident, nil
+		}
+	}
+
+	return nil, nil
+}
+
+
+// IdentFromFile gets all the Identifiers from a given BUILD file
+// *reads complete files only*
+func (a *Analyzer) IdentFromFile(uri lsp.DocumentURI, filecontent []string) ([]*Identifier,  error) {
+	filepath, err := GetPathFromURL(uri, "file")
+	if err != nil {
+		return nil, err
+	}
+	bytecontent, err := ioutil.ReadFile(filepath)
+	if err != nil {
+		return nil, err
+	}
+
+	stmt, err := a.parser.ParseData(bytecontent, filepath)
+	if err != nil {
+		return nil, err
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	var idents []*Identifier
+	for i, val := range stmt {
+		//fmt.Println(val.Ident.Action.Call)
+		if val.Ident != nil {
+			ident := &Identifier{
+				IdentStatement: val.Ident,
+				// -1 from asp.Statement.Pos.Line, as lsp position requires zero index
+				StartLine: val.Pos.Line - 1,
+			}
+			idents = append(idents, ident)
+
+			// fillin End for the previous iterated identifier
```

Comment:
don't think this is correct, the last identifier might have been any number of lines above.

Finding the end positions is Very Difficult though I think...